torch/distributions/transformed_distribution.py

```diff
@@ -76,5 +76,7 @@ def log_prob(self, value):
             x = transform.inv(y)
             log_prob -= transform.log_abs_det_jacobian(x, y)
             y = x
+        if self.base_dist._event_shape:
```

Comment:
Some transforms should already know their event shape. Also sometimes `log_prob` is just a number here (e.g. for `identity_transform`). I think it would be safer to do one of:
1. add an `event_shape` arg to pointwise transforms
2. implemement an `TensorTransform(base_transform, event_dim=1)` to wrap `ExpTransform`
3. make this hack a bit more robust:
```py
log_prob = ...sum of log_abs_det_jacobian_terms...
base_log_prob = self.base_dist.log_prob(y)  # knows the correct shape
if not isinstance(log_prob, numbers.Number):
    while log_prob.dim() > base_log_prob.dim():
        log_prob = log_prob.sum(-1)
log_prob += base_log_prob
```

Reply:
I like solution `1.` the best. I think fixing the hack as you did doesn't work because `log_prob` isn't a number most of the time - for example it can be `[sample_shape x batch_shape x ...]`. Let me rethink this and get back to you.

Reply:
:+1: We should also make `constraints` more aware of event dim. @tbrx is already doing this for multivariate normal in #52 and we'll need it e.g. for `AffineOperatorTransform` which transforms from `constraints.real_vector` to `constraints.real_vector`.

Reply:
After thinking about this a bit, I think a 4th approach could be minimally invasive:
- add a static `event_dim` attribute to each `Transform` (only about 4 lines of code diff)
- update the `log_prob` accumulation loop in `TransformedDistribution.log_prob()`

I'll sketch this in a PR so just to discuss